app/gui2/src/components/ComponentBrowser.vue

```diff
@@ -1,69 +1,108 @@
+    selectLastAfterRefresh()
+  }
+})
 
-const shown = ref(false)
-const scenePosition = ref(Vec2.Zero())
+// === Position ===
 
 const transform = computed(() => {
   const nav = props.navigator
-  const pos = scenePosition.value
+  const pos = props.position
   return `${nav.transform} translate(${pos.x}px, ${pos.y}px) scale(${
     1 / nav.scale
   }) translateY(-100%)`
 })
 
-function positionAtMouse(): boolean {
-  const nav = props.navigator
-  const mousePos = nav.sceneMousePos
-  if (mousePos == null) return false
-  scenePosition.value = mousePos
-  return true
+// === Input and Filtering ===
+
+const cbRoot = ref<HTMLElement>()
+const inputField = ref<HTMLElement>()
+const inputText = ref('')
+const filterFlags = ref({ showUnstable: false, showLocal: false })
+
+const currentFiltering = computed(() => {
+  let [qualifiedNamePattern, pattern] = qnSplit(inputText.value)
+  return new Filtering({ pattern, qualifiedNamePattern, ...filterFlags.value })
+})
+
+watch(currentFiltering, selectLastAfterRefresh)
+
+function handleDefocus(e: FocusEvent) {
+  const stillInside =
+    cbRoot.value != null &&
+    e.relatedTarget instanceof Node &&
+    cbRoot.value.contains(e.relatedTarget)
+  if (stillInside) {
+    if (inputField.value != null) {
+      inputField.value.focus({ preventScroll: true })
+    }
+  } else {
+    emit('finished')
+  }
 }
 
 // === Components List and Positions ===
 
-const componentStore = useComponentsStore()
+const suggestionDbStore = useSuggestionDbStore()
+
+const components = computed(() => {
+  return makeComponentList(suggestionDbStore.entries, currentFiltering.value)
+})
 
 const visibleComponents = computed(() => {
   if (scroller.value == null) return []
   const scrollPosition = animatedScrollPosition.value
-  const firstVisible = componentAtY(scrollPosition)
-  const lastVisible = componentAtY(animatedScrollPosition.value + scrollerSize.value.y)
-  return componentStore.components.slice(firstVisible, lastVisible + 1).map((component, i) => {
-    return { component, index: i + firstVisible }
+  const topmostVisible = componentAtY(scrollPosition)
+  const bottommostVisible = Math.max(
+    0,
+    componentAtY(animatedScrollPosition.value + scrollerSize.value.y),
+  )
+  return components.value.slice(bottommostVisible, topmostVisible + 1).map((component, i) => {
+    return { component, index: i + bottommostVisible }
   })
 })
 
 function componentPos(index: number) {
-  return index * ITEM_SIZE
+  return listContentHeight.value - (index + 1) * ITEM_SIZE
 }
 
 function componentAtY(pos: number) {
-  return Math.floor(pos / ITEM_SIZE)
+  return Math.floor((listContentHeight.value - pos) / ITEM_SIZE)
 }
 
 function componentStyle(index: number) {
   return { transform: `translateY(${componentPos(index)}px)` }
 }
 
 function componentColor(component: Component): string {
-  return componentStore.groups[component.group].color
+  return suggestionDbStore.groups[component.group ?? -1]?.color ?? '#006b8a'
```

Comment:
Instead of hardcoding the color in script, it could potentially be moved to css by returning `var(--default-component-color)`.

Reply:
This should be improved even more, because not each group has assigned color - this color pool for unspecified groups could also be in CSS, and I added a task to do it properly. I'll write a TODO here.